### ISSUE TITLE
Proposal: RBAC best-practices changes

### ISSUE THREAD

**Issue by omkensey** [kubernetes/helm#3320]
(Ref: RBAC comments in jetstack/cert-manager#229, Helm issue #3098)

In discussion of a Jetstack PR, I have come to believe we need to at least update the best practices doc for RBAC in charts, and possibly refactor those practices.

The bit that needs updating is the current assertion that "serviceAccountName is effectively a required value".  It should be optional, but generated to a unique name if not specified.

However, additional cases not handled by the current practice are:
* a service account already exists, and a new release should create RBAC or other resources that use it
* a service account should be created for an RBAC-enabled chart even when RBAC is not being used

In the Jetstack PR discussion, it was proposed that service account values should be split out from RBAC (since they are used by RBAC but exist independently of it) and created or not independently, e.g.:

```
rbac:
  create: true

serviceAccount:
  create: false
  serviceAccountName: "foo"
```

This adds a bit of complexity, but resolves the issue described above, and additionally allows charts to specify and use multiple service accounts if we add a layer under `serviceAccount` like so:

```
serviceAccounts:
  client:
    serviceAccountName: client-account
    create: false
  server:
    serviceAccountName: server-account
    create: true
```

My personal feeling is that the additional functionality, especially of the second form, is worth the small added complexity.

**Comment by technosophos** [comment]
I agree with this. I think we should actually alter the Tiller installer to do this. I don't _think_ this introduces any compatibility breakers. And we do this same thing with Brigade (add svc account by default, even if no RBAC), which seems to work for all users.

**Comment by unguiculus** [comment]
@omkensey I thought it through again and think the proposal makes sense. I updated the PR accordingly. Please have a look at this commit: https://github.com/jetstack/cert-manager/pull/229/commits/52f2c90b33a8857d2358ff99e75919e888ca9405

**Comment by unguiculus** [comment]
I went for a slightly shorter version:

```yaml
rbac:
  create: true

serviceAccount:
  create: false
  name: "foo"
```

**Comment by whereisaaron** [comment]
Our discussion on jetstack/cert-manager#229 has disappeared, but after some discussion there, I realized that conflating Service Accounts with RBAC was complicating the use cases. I'm strongly in favor of separating the two as in @unguiculus's and @omkensey's examples above.

**Comment by omkensey** [comment]
Per @unguiculus's example above, I think that's a good default for this (since probably a solid majority of charts that need service accounts only need one), and then we can have a standardized more-complex form (like mine :) ) for cases where one service account just won't do.  We don't want to open the floodgates but I think a pattern of "best practice for simple cases" and a more complex "best practice for other cases" would serve us well for this and some other uses.

**Comment by unguiculus** [comment]
@whereisaaron The discussion is still there. Github just hides it as outdated.

**Comment by depohmel** [comment]
Guys, I looked thru all discussions which were mentions here, but I didn't find why serviceAccount is on the root level of values and not under RBAC? Even that most of the clusters have RBAC enabled by default. But sometimes people are setting RBAC off cluster for testing purpose. And in those environments, SA creation might fail. So it looks like for rbac.enable=false, chart should not even try to create SA.

**Comment by whereisaaron** [comment]
Hi @depohmel. Service Account creation won't fail even if you don't have or use RBAC. Service Accounts know nothing of RBAC, and don't care whether you have it or not 😄 

This is because Service Accounts are not part of RBAC. They are a separate function that exists independent of RBAC. RBAC is something you optionally add on top, to tell the k8s API to limit access to the API for certain Service Accounts. 

You still need Service Accounts, even you don't use RBAC (or you use ABAC or some future RBACv2 instead).

**Comment by depohmel** [comment]
Oohhhh. Thanks. This was a missing part. Now it does make sense.

**Comment by unguiculus** [comment]
Fixed in #3408.

**Comment by whereisaaron** [comment]
good stuff!

...aaron boodles off to update all his old charts... 😢
